docs/book/src/developer/providers/contracts/infra-machine.md

```diff
@@ -1,118 +1,486 @@
+    // but instead indicate that something is fundamentally wrong with the FooMachine and that it cannot be recovered.
+    // +optional
+    FailureReason *capierrors.ClusterStatusError `json:"failureReason,omitempty"`
+    
+    // FailureMessage will be set in the event that there is a terminal problem reconciling the FooMachine
+    // and will contain a more verbose string suitable for logging and human consumption.
+    //
+    // This field should not be set for transitive errors that can be fixed automatically or with manual intervention,
+    // but instead indicate that something is fundamentally wrong with the FooMachine and that it cannot be recovered.
+    // +optional
+    FailureMessage *string `json:"failureMessage,omitempty"`
+    
+    // See other rules for more details about mandatory/optional fields in InfraMachine status.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+```
+
+Once `status.failureReason` and `status.failureMessage` are set on the InfraMachine resource, the Machine "core" controller
+will bubble up those info in the corresponding fields in Machine's `status`.
+
+Please note that once failureReason/failureMessage is set in Machine's `status`, the only way to recover is to delete and
+recreate the Machine (it is a terminal failure).
+
+<aside class="note warning">
+
+<h1>Heads up! this will change with the v1beta2 contract</h1>
+
+When the v1beta2 contract will be released (tentative Apr 2025), support for `status.failureReason` and `status.failureMessage`
+will be dropped.
+
+See [Improving status in CAPI resources].
+
+</aside>
+
+### InfraMachineTemplate, InfraMachineTemplateList resource definition
+
+For a given InfraMachine resource, you should also add a corresponding InfraMachineTemplate resources in order to use it in ClusterClasses.
+The template resource MUST be named as `<InfraMachine>Template`.
+
+```go
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=foomachinetemplates,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// FooMachineTemplate is the Schema for the foomachinetemplates API.
+type FooMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec FooMachineTemplateSpec `json:"spec,omitempty"`
+}
+
+type FooMachineTemplateSpec struct {
+    Template FooMachineTemplateResource `json:"template"`
+}
+
+type FooMachineTemplateResource struct {
+    // Standard object's metadata.
+    // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+    // +optional
+    ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
+    Spec FooMachineSpec `json:"spec"`
+}
+```
+
+NOTE: in this example InfraMachineTemplate's `spec.template.spec` embeds `FooMachineSpec` from InfraMachine. This might not always be
+the best choice depending of if/how InfraMachine's spec fields applies to many clusters vs only one.
+
```

Comment:
```suggestion
```

Reply:
Why dropping this, theres is a nit (s/clusters/machines) but it still makes sense IMO

Reply:
After thinking about some more, yup I think you're right